File: EgoCapture/Models/TaskVerbsCatalog.swift
import Foundation

/// Bundle mirror of `Resources/task_verbs.json` — per-`task_category_code`
/// verb lists in PT / EN / ES (parallel indices).
struct TaskVerbsBundle: Codable {
    let schemaVersion: String
    let generatedAt: String
    let entries: [Entry]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case generatedAt = "generated_at"
        case entries
    }

    struct Entry: Codable, Hashable {
        let taskCategoryCode: String
        let verbsPt: [String]
        let verbsEn: [String]
        let verbsEs: [String]

        enum CodingKeys: String, CodingKey {
            case taskCategoryCode = "task_category_code"
            case verbsPt = "verbs_pt"
            case verbsEn = "verbs_en"
            case verbsEs = "verbs_es"
        }
    }
}

enum TaskVerbsLoader {
    /// Lazily-loaded index keyed by `task_category_code`.
    static let shared: [String: TaskVerbsBundle.Entry] = {
        guard let url = Bundle.main.url(forResource: "task_verbs", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            print("[TaskVerbsLoader] task_verbs.json missing from bundle.")
            return [:]
        }
        do {
            let bundle = try JSONDecoder().decode(TaskVerbsBundle.self, from: data)
            return Dictionary(uniqueKeysWithValues: bundle.entries.map { ($0.taskCategoryCode, $0) })
        } catch {
            print("[TaskVerbsLoader] Decode failed: \(error)")
            return [:]
        }
    }()

    static func verbs(for taskCategoryCode: String) -> TaskVerbsBundle.Entry? {
        shared[taskCategoryCode]
    }
}

// MARK: - Display casing

extension String {
    /// Uppercases only the first character of a verb token for on-screen use.
    /// Underlying `task_verbs.json` and `taxonomy.json` stay lowercased.
    var verbDisplayCased: String {
        guard !isEmpty else { return self }
        return prefix(1).uppercased() + dropFirst()
    }
}

File: android/app/src/main/java/com/kgeneye/eye/taxonomy/TaskVerbsCatalog.kt
package com.kgeneye.eye.taxonomy

import android.content.Context
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.Locale

/**
 * Kotlin mirror of `task_verbs.json` — per `task_category_code` verb lists in
 * PT / EN / ES. Indices are parallel across the three arrays, so the selected
 * set is simply a set of indices; the wizard then projects it to three
 * localised arrays when persisting the session taxonomy.
 */
@Serializable
data class TaskVerbsBundle(
    @SerialName("schema_version") val schemaVersion: String,
    @SerialName("generated_at") val generatedAt: String,
    val entries: List<Entry>,
) {
    @Serializable
    data class Entry(
        @SerialName("task_category_code") val taskCategoryCode: String,
        @SerialName("verbs_pt") val verbsPt: List<String>,
        @SerialName("verbs_en") val verbsEn: List<String>,
        @SerialName("verbs_es") val verbsEs: List<String>,
    )
}

object TaskVerbsLoader {
    private const val ASSET_NAME = "task_verbs.json"

    @Volatile private var cached: Map<String, TaskVerbsBundle.Entry>? = null

    @OptIn(ExperimentalSerializationApi::class)
    fun load(context: Context): Map<String, TaskVerbsBundle.Entry> {
        cached?.let { return it }
        synchronized(this) {
            cached?.let { return it }
            val json = Json {
                ignoreUnknownKeys = true
                isLenient = true
            }
            val data = context.assets.open(ASSET_NAME).bufferedReader().use { it.readText() }
            val bundle = json.decodeFromString(TaskVerbsBundle.serializer(), data)
            val index = bundle.entries.associateBy { it.taskCategoryCode }
            cached = index
            return index
        }
    }

    fun verbs(context: Context, taskCategoryCode: String): TaskVerbsBundle.Entry? =
        load(context)[taskCategoryCode]
}

/** Uppercases only the first character of a verb token for on-screen use. */
fun String.verbDisplayCased(): String =
    if (isEmpty()) this else substring(0, 1).uppercase() + substring(1)

/** Returns the localised verb list for the current locale (falls back to EN). */
fun TaskVerbsBundle.Entry.localizedList(): List<String> {
    val lang = Locale.getDefault().language.lowercase()
    return when {
        lang.startsWith("pt") -> verbsPt
        lang.startsWith("es") -> verbsEs
        else -> verbsEn
    }
}
